packages/core/src/Synchronizer.ts

```diff
@@ -176,80 +215,68 @@ export class Synchronizer extends EventEmitter {
         this.settings.fieldCount = config.fieldCount
         this.settings.sumFieldCount = config.sumFieldCount
 
-        await this.findStartBlock()
-
-        if (this.attesterId === BigInt(0)) return
-
-        // load the GST for the current epoch
-        // assume we're resuming a sync using the same database
-        const epochs = await this._db.findMany('Epoch', {
-            where: {
-                attesterId: this.attesterId.toString(),
-                sealed: false,
-            },
-        })
-        this.defaultStateTreeLeaf = BigInt(0)
-        this.stateTree = new IncrementalMerkleTree(
-            this.settings.stateTreeDepth,
-            this.defaultStateTreeLeaf
-        )
-        // if it's a new sync, start with epoch 0
-        const epoch = epochs[epochs.length - 1]?.number ?? 0
-        // otherwise load the leaves and insert them
-        // TODO: index consistency verification, ensure that indexes are
-        // sequential and no entries are skipped, e.g. 1,2,3,5,6,7
-        const leaves = await this._db.findMany('StateTreeLeaf', {
-            where: {
-                epoch,
-                attesterId: this.attesterId.toString(),
-            },
-            orderBy: {
-                index: 'asc',
-            },
-        })
-        for (const leaf of leaves) {
-            this._stateTree.insert(leaf.hash)
-        }
+        await this._findStartBlock()
+        this.setupComplete = true
     }
 
-    async findStartBlock() {
+    private async _findStartBlock() {
         // look for the first attesterSignUp event
         // no events could be emitted before this
-        const filter = this.unirepContract.filters.AttesterSignedUp(
-            this.attesterId
-        )
-        const events = await this.unirepContract.queryFilter(filter)
-        if (events.length === 0) {
-            throw new Error(
-                '@unirep/core:Synchronizer: failed to fetch genesis event'
-            )
+        let events: ethers.Event[] = []
+        if (this._syncAll) {
+            const filter = this.unirepContract.filters.AttesterSignedUp()
+            events = await this.unirepContract.queryFilter(filter)
+        } else if (this._attesterId.length) {
+            for (let id of this._attesterId) {
```

Comment:
Probably don't do this in a synchronous `for` loop. You should be able to batch these requests or construct a single filter with an array of `attesterId`.

Reply:
[fe55a01](https://github.com/Unirep/Unirep/pull/367/commits/fe55a01161d77790111ced37cb26ebb5c6e86520)